### PR TITLE
Core Algorithms delicate-cow

### PR DESCRIPTION
**Pull request by TrevorJamesH**

### CODE REVIEW

File: src/binarySearch.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import dataTypeCheck from '../src/dataTypeCheck'
+
+export default function binarySearch(array, targetValue) {
+
+	if(!dataTypeCheck(array,'array')) {
```
Comment (mantinone) [inline src/binarySearch.js:5]: Nice thinking having a function for this.

File: test/binarySearch_test.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+  it('should search for a number in an array and return its index', function() {
+    const numbers = [5, 10, 15, 20, 25, 30, 35, 40, 45, 50]
+    expect(binarySearch(numbers, 45)).to.equal(8)
+  })
+
+  it('should search for a number in an array and return its index', function() {
+    const numbers = [5, 10, 15]
+    expect(binarySearch(numbers, 5)).to.equal(0)
+  })
```
Comment (DianaVashti) [inline test/binarySearch_test.js:21]: This test is the same (and has the same description) as the test above it. Omit tests that test the exact same thing.

File: src/binarySearch.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  while(max > min) {
+  	guess = Math.round((max + min) / 2);
+    if(array[guess] === targetValue){
+      return guess
+    }
+    if(array[guess] < targetValue) {
+      min = guess + .5
+    }
+    else if(array[guess] > targetValue) {
+      max = guess - .5
```
Comment (mantinone) [inline src/binarySearch.js:20]: Interesting use of a fractional number to avoid having to do off-by-one checks.

File: test/binarySearch_test.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+  it('should search for a number in an array, return false if its not there', function() {
+    const numbers = [1, 2, 3, 4]
+    expect(binarySearch(numbers, 5)).to.equal(false)
+  })
```
Comment (mantinone) [inline test/binarySearch_test.js:26]: To really thoroughly test this algorithm, I would recommend searching for numbers larger than and smaller than any other number in the array, as well as searching for the first and last items.

File: test/bubbleSort_test.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+describe('bubbleSort()', function() {
+  it('should be a function', function() {
+    expect(bubbleSort).to.be.a('function')
+  })
+
+  it('should sort through indices of an array', function() {
+    expect(bubbleSort([4,3,2,1])).to.eql([1,2,3,4])
+    expect(bubbleSort([10,2,7,5,8,3,6,1,4,9])).to.eql([1,2,3,4,5,6,7,8,9,10])
+    expect(bubbleSort([10])).to.eql([10])
```
Comment (DianaVashti) [inline test/bubbleSort_test.js:12]: This third `expect` is testing for what happens if the input is only one index. This should be a separate test with a description like: `Returns array as-is if array is one index long`

File: test/bubbleSortRecursion_test.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+      expect(bubbleSort([456,9027,98])).to.eql([98,456,9027])
+    })
+  })
+
+  context("when given [1,2,3,4,5,6,7,8,9,10,9,8,7,6,5,4,3,2,1]", function(){
+    it("should return [1,1,2,2,3,3,4,4,5,5,6,6,7,7,8,8,9,9,10]", function(){
+      expect(bubbleSort([1,2,3,4,5,6,7,8,9,10,9,8,7,6,5,4,3,2,1])).to.eql([1,1,2,2,3,3,4,4,5,5,6,6,7,7,8,8,9,9,10])
+    })
+  })
+})
```
Comment (DianaVashti) [inline test/bubbleSortRecursion_test.js:33]: Each of these should have a more description description about what aspect of behavior the algorithm is proving or disproving. See above.

File: test/bubbleSortRecursion_test.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+  context("when given [5,5,5,9,105,5,5,1]", function() {
+    it("should return [1,5,5,5,5,5,9,105]", function() {
+      expect(bubbleSort([5,5,5,9,105,5,5,1])).to.eql([1,5,5,5,5,5,9,105])
+    })
+  })
+
+  context("when given [456,9027,98]", function() {
+    it("should return [98,456,9027]", function() {
+      expect(bubbleSort([456,9027,98])).to.eql([98,456,9027])
```
Comment (DianaVashti) [inline test/bubbleSortRecursion_test.js:24]: Is this testing for functionality when inputs are of varying sizes? State, if so.

File: test/bubbleSortRecursion_test.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+  context("when given [9,4,6,8,7,5,1,3,2]", function() {
+    it("should return [1,2,3,4,5,6,7,8,9]", function() {
+      expect(bubbleSort([9,4,6,8,7,5,1,3,2])).to.eql([1,2,3,4,5,6,7,8,9])
+    })
+  })
+
+  context("when given [5,5,5,9,105,5,5,1]", function() {
+    it("should return [1,5,5,5,5,5,9,105]", function() {
+      expect(bubbleSort([5,5,5,9,105,5,5,1])).to.eql([1,5,5,5,5,5,9,105])
```
Comment (DianaVashti) [inline test/bubbleSortRecursion_test.js:18]: This should say something like: `Should return sorted array even if duplicate values are given`

File: test/bubbleSortRecursion_test.js [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+describe('bubbleSort()', function() {
+
+  it('should be a function', function() {
+    expect(bubbleSort).to.be.a('function')
+  })
+
+  context("when given [9,4,6,8,7,5,1,3,2]", function() {
+    it("should return [1,2,3,4,5,6,7,8,9]", function() {
+      expect(bubbleSort([9,4,6,8,7,5,1,3,2])).to.eql([1,2,3,4,5,6,7,8,9])
```
Comment (DianaVashti) [inline test/bubbleSortRecursion_test.js:12]: This should say something like: `Returns array sorted numerically: lowest to highest`

**Review by DianaVashti — COMMENTED** [review]
Creative implementation of bubblesorts.

File: src/collatz.js [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+        number = number * 3 + 1
+        collatzArray.push(number)
+      }
+      else {
+        number = number / 2
+        collatzArray.push(number)
+      }
+    }
+    return collatzArray
+  }
```
Comment (DianaVashti) [inline src/collatz.js:22]: Nice work.

**Review by DianaVashti — COMMENTED** [review]
Review (part 2): The code for the src files are looking good. Code is clean, variable names are description, and the logic is sound. Make sure your test descriptions are specific to the precise thing you are testing for & that there are not duplicate redundant tests.

File: src/connectedGraph.js [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+  let toCheck = [allPoints[0]]
+
+  while(toCheck.length>0) {
+    let neighbors = graph[toCheck[0]]
+    neighbors.map((point) => toCheck.push(point))
+    connected.push(toCheck.shift())
+    toCheck = setCompliment(connected,toCheck)
+  }
+
+  return setCompliment(connected,allPoints).length === 0
```
Comment (mantinone) [inline src/connectedGraph.js:20]: Good thinking to use your old functions, but this is more complicated than you need to be to check this.

All you need to do is check if connected contains all points.

File: src/dataTypeCheck.js [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+    return true
+  }
+  if(expectedType === 'string' && typeof data === 'string') {
+    return true
+  }
+  if(expectedType === 'number' && !isNaN(data)) {
+    return true
+  }
+  return false
+}
```
Comment (mantinone) [inline src/dataTypeCheck.js:15]: I would recommend using an Object-literal lookup table for this.  
{ 'array': () => { return Array.isArray(data)} }...etc

**Review by mantinone — COMMENTED** [review]
I didn't have time to get to everything, but it looks good!  I might come back later and add more comments.

**Pull request closed without merging**